### ISSUE TITLE
One-way encrypted file system

### ISSUE THREAD

**Issue by bep** [spf13/afero#61]
Seems like a good fit for afero, ala:

http://owefs.firelet.net/

**Comment by xor-gate** [comment]
It seems it has been removed from github and the site is down. Only google has something in the cache.

**Comment by mbertschler** [comment]
I read something about that topic. Do I understand it right, that you would have an `EncryptFs` that can only write files and then a `DecryptFs` that can only read (but not write) the files back?

**Comment by bep** [comment]
Some more info here: https://www.reddit.com/r/linux/comments/418oat/owefs_oneway_encrypted_filesystem/

I think I understand why the author "bailed out" -- encryption stuff is incredible hard to get just right.

**Comment by xor-gate** [comment]
Yes you are right:
* Public/privatekey mechanism
* Per-file random key
* Encrypt file with random key, append encrypted random key

**Comment by mbertschler** [comment]
I can understand that too, and we would also need to be very careful implementing something like that for afero, although I really would like something like that.

Why is the key (I suppose the public key used to encrypt the file) appended to the file? You would anyways need some extra place to store the keys to decrypt the files, this could also be a list of path > private key pairs. Just wondering...

**Comment by xor-gate** [comment]
Yeah you are right, appending sucks really hard. You could hash the encrypted file and then use the checksum to store the key. Then you would not need an key-value store.

**Comment by mbertschler** [comment]
I was thinking about using a [boltDB](https://github.com/boltdb/bolt) store, which would then be a single file containing all the needed keys. The checksum would also be good to check the file integrity before decrypting it.

**Comment by xor-gate** [comment]
Yeah I was also thinking about that, but when you want to decrypt with a set of bash (openssl) scripts then with boltDB this is not possible. It is ugly, but storing all the keys in bolt for long term feels not so good.

**Comment by mbertschler** [comment]
Okay that of course would kill interoperability. 
The requirements are too vague for now, I think we would need to think about what usecases such a filesystem should support before planning something that can be implemented.

I am currently building a backup/sync program that could also benefit from encrypted storage, although for that I would be okay for using just one key for all files and I also think that one way encryption wouldn't benefit me.

Any other ideas or examples?
